### PR TITLE
use `textwidth` in Printf for `%s` and `%c` widths

### LINKED ISSUE #41068: @printf should use textwidth for "%ns" and "%0.ns"
This seems wrong:
```jl
julia> @printf("%5s", "foo")
  foo

julia> @printf("%5s", "föó")
föó

julia> @printf("%.2s", "föó")
fo
```
It seems like `@printf` (in [this code](https://github.com/JuliaLang/julia/blob/9f3265399227fbfc4f0160ec3592a9262bd3eb5f/stdlib/Printf/src/Printf.jl#L239-L277)) should really use `textwidth` to compute string widths for formatting purposes, similar to this [discussion on discourse](https://discourse.julialang.org/t/indexing-unicode-strings/62325/4?u=stevengj).

(To improve backward compatibility, I suppose we could use `textwidth` only for `%S`, which is currently a rarely used synonym for `%s`, or use another format code entirely.)

### PR DESCRIPTION
**Pull request by stevengj**
Closes #41068 (similar to the discussion regarding #38256).

### CODE REVIEW

**Review by quinnj — APPROVED** [review]
Thanks!

**Comment by stevengj** [comment]
Not sure what is going on with Documenter.jl:
```
��� Expected output:
��� 
��� 0 0.0 -0.007812
��� 
��� !!! compat "Julia 1.7"
���     Starting in Julia 1.7, `%s` (string) and `%c` (character) widths are computed
���     using [`textwidth`](@ref), which e.g. ignores zero-width characters
���     (such as combining characters for diacritical marks) and treats certain
���     "wide" characters (e.g. emoji) as width `2`.
��� 
���   diff =
���    0 0.0 -0.007812
���    
���    !!! compat "Julia 1.7"
���        Starting in Julia 1.7, `%s` (string) and `%c` (character) widths are computed
���        using [`textwidth`](@ref), which e.g. ignores zero-width characters
���        (such as combining characters for diacritical marks) and treats certain
���        "wide" characters (e.g. emoji) as width `2`.-0.007812
��� @ Documenter.DocTests /buildworker/worker/doctest_linux64/build/doc/deps/packages/Documenter/bFHi4/src/DocTests.jl:373
ERROR: LoadError: `makedocs` encountered a doctest error. Terminating build
```

File: stdlib/Printf/src/Printf.jl [hunk 1]
```diff
@@ -858,6 +865,12 @@ Inf Inf NaN NaN
 
 julia> @printf "%.0f %.1f %f" 0.5 0.025 -0.0078125
 0 0.0 -0.007812
+
+!!! compat "Julia 1.7"
+    Starting in Julia 1.7, `%s` (string) and `%c` (character) widths are computed
+    using [`textwidth`](@ref), which e.g. ignores zero-width characters
+    (such as combining characters for diacritical marks) and treats certain
+    "wide" characters (e.g. emoji) as width `2`.
 ```
```
Comment (fredrikekre) [inline stdlib/Printf/src/Printf.jl:874]:
```suggestion
```

!!! compat "Julia 1.7"
    Starting in Julia 1.7, `%s` (string) and `%c` (character) widths are computed
    using [`textwidth`](@ref), which e.g. ignores zero-width characters
    (such as combining characters for diacritical marks) and treats certain
    "wide" characters (e.g. emoji) as width `2`.
```

**Comment by stevengj** [comment]
Hooray, tests are green.  Should be good to merge?

**Comment by stevengj** [comment]
For some reason, this PR seems to be missing from the Julia 1.7 release — any idea why?

**Pull request merged**